syntax = "proto3";

package emotalk;

// Service định nghĩa cho EmoTalk API
service EmoTalkService {
  // Stream blendshapes từ audio input
  rpc ProcessAudio(AudioRequest) returns (stream BlendshapeResponse);
  
  // Health check
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// Request chứa audio data
message AudioRequest {
  string request_id = 1;           // ID unique cho request
  bytes audio_data = 2;            // Audio data dạng bytes (WAV format, 16kHz)
  int32 emotion_level = 3;         // Emotion level: 0 or 1
  int32 person_id = 4;             // Person ID: 0-23
  bool post_processing = 5;        // Có sử dụng post-processing (smoothing, blinking) không
  float chunk_duration = 6;        // Thời lượng mỗi chunk (seconds), default 25.0
  float chunk_overlap = 7;         // Thời lượng overlap giữa các chunks (seconds), default 0.5
}

// Response stream chứa blendshapes
message BlendshapeResponse {
  string request_id = 1;           // ID của request tương ứng
  int32 chunk_index = 2;           // Index của chunk hiện tại
  repeated BlendshapeFrame frames = 3;  // Danh sách các frames trong chunk
  bool is_final = 4;               // Có phải chunk cuối cùng không
  string error_message = 5;        // Thông báo lỗi nếu có
}

// Frame chứa 52 blendshapes và timecode
message BlendshapeFrame {
  double timecode = 1;             // Timecode của frame (seconds)
  int32 frame_number = 2;          // Số thứ tự frame
  repeated float blendshapes = 3;  // 52 giá trị blendshapes
}

// Health check messages
message HealthCheckRequest {
  string service = 1;
}

message HealthCheckResponse {
  enum ServingStatus {
    UNKNOWN = 0;
    SERVING = 1;
    NOT_SERVING = 2;
  }
  ServingStatus status = 1;
}
